subsys/net/ip/utils.c

```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
 
 	return false;
 }
+
+#if defined(CONFIG_NET_IPV6) || defined(CONFIG_NET_IPV4)
+static bool convert_port(const char *buf, u16_t *port)
+{
+	char *endptr;
+	int tmp;
+
+	tmp = strtol(buf, &endptr, 10);
+	if ((endptr == buf && tmp == 0) ||
+	    !(*buf != '\0' && *endptr == '\0') ||
+	    tmp > 0xffff || tmp < 0) {
+		return false;
+	}
+
+	*port = tmp;
+
+	return true;
+}
+#endif /* CONFIG_NET_IPV6 || CONFIG_NET_IPV4 */
+
+bool net_ipaddr_parse(const char *str, size_t str_len, struct sockaddr *addr)
+{
+	int count, i, ret;
+	u16_t port;
+	char *ptr;
+
+	if (*str == '[') {
+#if defined(CONFIG_NET_IPV6)
+		/* IPv6 address with port number */
+		struct in6_addr *addr6;
+		char ipaddr[INET6_ADDRSTRLEN + 1];
+		int end, len;
+
+		ptr = strstr(str, "]");
+		if (!ptr || (ptr >= (str + str_len))) {
+			return false;
+		}
+
+		len = min(INET6_ADDRSTRLEN, str_len);
+		end = min(len, ptr - (str + 1));
+
+		memcpy(ipaddr, str + 1, end);
+		ipaddr[end] = '\0';
+
+		addr6 = &net_sin6(addr)->sin6_addr;
+
+		ret = net_addr_pton(AF_INET6, ipaddr, addr6);
+		if (ret < 0) {
+			return false;
+		}
+
+		net_sin6(addr)->sin6_family = AF_INET6;
+
+		if ((ptr + 1) < (str + str_len) && *(ptr + 1) == ':') {
+			memcpy(ipaddr, ptr + 2, str_len - end);
+			ipaddr[str_len - end] = '\0';
+
+			ret = convert_port(ipaddr, &port);
+			if (!ret) {
+				return false;
+			}
+
+			net_sin6(addr)->sin6_port = htons(port);
+
+			NET_DBG("IPv6 server %s port %d",
+				net_addr_ntop(AF_INET6, addr, ipaddr,
+					      sizeof(ipaddr) - 1), port);
+		} else {
+			NET_DBG("IPv6 server %s",
+				net_addr_ntop(AF_INET6, addr, ipaddr,
+					      sizeof(ipaddr) - 1));
+		}
+
+		return true;
+#else
+		return false;
+#endif /* CONFIG_NET_IPV6 */
+	}
+
+	count = i = 0;
+	while (str[i]) {
```

Comment:
Yes, and "for (count = i = 0; str[i] && i < str_len; i++) {" is even better as it checks the string length too.